src/aspire/abinitio/commonline_irls.py

```diff
@@ -0,0 +1,317 @@
+import logging
+
+import numpy as np
+from scipy.sparse.linalg import eigsh
+
+from aspire.abinitio import CommonlineLUD
+from aspire.utils import trange
+
+logger = logging.getLogger(__name__)
+
+
+class CommonlineIRLS(CommonlineLUD):
+    """
+    Estimate 3D orientations using Iteratively Reweighted Least Squares
+    (IRLS) as described in the following publication:
+    L. Wang, A. Singer, and  Z. Wen, Orientation Determination of Cryo-EM Images Using
+    Least Unsquared Deviations, SIAM J. Imaging Sciences, 6, 2450-2483 (2013).
+    """
+
+    def __init__(
+        self,
+        src,
+        *args,
+        num_itrs=10,
+        eps_weighting=1e-3,
+        alpha=None,
+        max_rankZ=None,
+        max_rankW=None,
+        **kwargs,
+    ):
+        """
+        Initialize a class for estimating 3D orientations using an IRLS-based optimization.
+        See CommonlineLUD for additional arguments.
+
+        :param num_itrs: Number of iterations for iterative reweighting. Default is 10.
+        :param eps_weighting: Regularization value for reweighting factor. Default is 1e-3.
+        :param alpha: Spectral norm constraint for IRLS algorithm. Default is None, which
+            does not apply a spectral norm constraint. To apply a spectral norm constraint provide
+            a value in the range [2/3, 1), 2/3 is recommended.
+        :param max_rankZ: Maximum rank used for projecting the Z matrix (for adaptive projection).
+            If None, defaults to max(6, n_img // 4).
+        :param max_rankW: Maximum rank used for projecting the W matrix (for adaptive projection).
+            If None, defaults to max(6, n_img // 4).
+        """
+
+        self.num_itrs = num_itrs
+        self.eps_weighting = eps_weighting
+
+        # Adjust rank limits
+        max_rankZ = max_rankZ or max(6, src.n // 4)
+        max_rankW = max_rankW or max(6, src.n // 4)
+
+        super().__init__(
+            src,
+            *args,
+            max_rankZ=max_rankZ,
+            max_rankW=max_rankW,
+            alpha=alpha,
+            **kwargs,
+        )
+
+    def estimate_rotations(self):
+        """
+        Estimate rotation matrices using the common lines method with IRLS optimization.
+        """
+        logger.info("Computing the common lines matrix.")
+        self.build_clmatrix()
+
+        self.S = self._construct_S(self.clmatrix)
+        weights = np.ones(2 * self.n_img, dtype=self.dtype)
+        gram = np.eye(2 * self.n_img, dtype=self.dtype)
+        if self.alpha is None:
+            A, b = self._sdp_prep()
+            for _ in trange(self.num_itrs, desc="Performing iterative re-weighting."):
+                S_weighted = weights * self.S
+                gram = self._compute_gram_matrix(S_weighted, A, b)
```

Comment:
What's the difference between `_compute_gram_matrix` and `_compute_Gram`? Why is one used with spectral norm constraint and one without?

Reply:
`_compute_gram_matrix` is calling the SDP solver in `CommonlineSDP` and `_compute_Gram` is calling the ADMM implementation in this file.

Maybe these should be renamed to indicate this. What do you think of `_compute_gram_SDP` and `_compute_gram_IRLS`?

Reply:
Updated!